src/lib/zork/zork.ts

```diff
@@ -191,6 +173,108 @@ const handleCmdGet = (ctx: Context) => {
+        ctx.log(`decremented numGetters to ${numGetters}`);
+      } else if (ctx.mode === 'get') {
+        numGivers--;
+        ctx.log(`decremented numGivers to ${numGivers}`);
+      }
+    }
+  };
+  conn.onicecandidate = (event: any) => ctx.reply(JSON.stringify(event));
+  conn.ondatachannel = (event: any) => onDataChannel(ctx, event.channel);
+};
+
+const initSocksServer = (ctx: Context) => {
+  // If we're already getting access from a peer, we already have a SOCKS
+  // server running on SOCKS_PORT, so use 0 so OS grabs us a free port.
+  const port = numGivers ? 0 : SOCKS_PORT;
+  ctx.log(`starting local SOCKS server on port ${port ? port : '[TBD]'}`);
+  const server = ctx.socksServer = new SocksServer(SOCKS_HOST, port);
+  server.onConnection((sessionId) => onSocksConnection(ctx, sessionId));
+  server.listen().then(() => {
+    const port = server.address().port;
+    ctx.log(`[socksServer] listening on ${SOCKS_HOST}:${port}`);
+    ctx.log(`[socksServer] Test with e.g. curl -x socks5h://${SOCKS_HOST}:${port} httpbin.org/ip`);
+  });
+};
+
+const onDataChannel = (ctx: Context, channel: any) => {
+  const channelId = channel.label;
+  ctx.log(`onDataChannel: [channel ${channelId}] [channel.readyState ${channel.readyState}]`);
+
+  channel.onclose = () => ctx.log(`[channel ${channelId}] datachannel closed`);
+  channel.onerror = (err: any) => ctx.log.error(`[channel ${channelId}] datachannel error: ${err}`);
+  if (ctx.mode === 'get' || channelId === BOOTSTRAP_CHANNEL_ID) {
+    // ondatachannel events don't fire for the peer creating the datachannel.
+    // Since the getter is the one that creates datachannels, we don't expect
+    // this to run when we're the getter, so just close the datachannel. Also
+    // close it if it's the bootstrap datachannel, as it's served its purpose.
+    channel.onopen = () => channel.close();
+  } else {
+    channel.onopen = () => ctx.log(`[channel ${channelId}] datachannel opened`);
+    initSocksSession(ctx, channel);
+    channel.onmessage = (event: any) => ctx.socksSession.handleDataFromSocksClient(event.data);
+  }
+};
+
+const onSocksConnection = (ctx: Context, sessionId: any) => {
+  ctx.log(`[socksServer] new connection: ${sessionId}`);
+  const channel = ctx.rtc.conn.createDataChannel(sessionId);
+  return {
+    // SOCKS client -> datachannel
+    handleDataFromSocksClient: (bytes: ArrayBuffer) => channel.send(bytes),
+    // SOCKS client <- datachannel
+    onDataForSocksClient: (callback: (buffer: ArrayBuffer) => void) => {
+      channel.onmessage = (event: any) => callback(event.data);
+      return this;
+    },
+    handleDisconnect: () => {
+      channel.close();
+      ctx.log(`[socksServer] [session ${sessionId}] client disconnect, closed datachannel`);
+    },
+    onDisconnect: (callback: () => void) => {
+      // This callback calls client.end(), so we ignore it.
+      ctx.log(`[socksServer] [session ${sessionId}] client disconnected`);
+      return this;
+    }
```

Comment:
(I'm not sure I fully get the intended functions and differences between handleDisconnect and onDisconnect, so sorry if this is a dumb question.)